File: model_upgrade/heuristics.py
import random
import time

from model_upgrade.bitsets import (
    MAX_BITSET_NODES,
    mask_degree,
    mask_intersection,
    mask_vertices,
    neighbor_sets_to_masks,
)


def _iter_bits(mask: int):
    while mask:
        low = mask & -mask
        yield low.bit_length() - 1
        mask &= ~low


def greedy_clique(
    neighbor_sets: list[set[int]],
    vertex_order: list[int],
) -> list[int]:
    """Build a clique by scanning vertices in the given order."""
    clique: list[int] = []
    clique_set: set[int] = set()
    for vertex in vertex_order:
        if not clique_set or clique_set.issubset(neighbor_sets[vertex]):
            clique.append(vertex)
            clique_set.add(vertex)
    return clique


def greedy_clique_grasp(
    neighbor_sets: list[set[int]],
    rng: random.Random,
    start_vertex: int | None = None,
    rcl_size: int = 4,
    neighbor_masks: list[int] | None = None,
) -> list[int]:
    """GRASP: repeatedly pick from the top-degree restricted candidate list."""
    n = len(neighbor_sets)
    use_masks = neighbor_masks is not None and n <= MAX_BITSET_NODES

    clique: list[int] = []
    if start_vertex is not None:
        clique = [start_vertex]

    if use_masks:
        assert neighbor_masks is not None
        if not clique:
            candidates_mask = (1 << n) - 1
        else:
            candidates_mask = mask_intersection(*[neighbor_masks[v] for v in clique])
            for v in clique:
                candidates_mask &= ~(1 << v)

        while candidates_mask:
            candidates = mask_vertices(candidates_mask)
            degrees = [(v, mask_degree(neighbor_masks[v])) for v in candidates]
            degrees.sort(key=lambda item: item[1], reverse=True)
            top = [v for v, _ in degrees[: min(rcl_size, len(degrees))]]
            chosen = rng.choice(top)
            clique.append(chosen)
            candidates_mask &= neighbor_masks[chosen]
            candidates_mask &= ~(1 << chosen)
        return clique

    clique_set = set(clique)
    if not clique_set:
        candidates = set(range(n))
    else:
        candidates = set.intersection(*(neighbor_sets[v] for v in clique_set))

    while candidates:
        ranked = sorted(candidates, key=lambda v: len(neighbor_sets[v]), reverse=True)
        top = ranked[: min(rcl_size, len(ranked))]
        chosen = rng.choice(top)
        clique.append(chosen)
        clique_set.add(chosen)
        candidates = candidates & neighbor_sets[chosen]
        candidates -= clique_set

    return clique


def random_restarts(
    neighbor_sets: list[set[int]],
    deadline: float,
    rng: random.Random,
    degeneracy: list[int] | None = None,
    neighbor_masks: list[int] | None = None,
) -> list[int]:
    """Run greedy construction with varied vertex orderings until the deadline."""
    n = len(neighbor_sets)
    degrees = [len(neighbors) for neighbors in neighbor_sets]
    best: list[int] = []

    static_orders = [
        sorted(range(n), key=lambda v: degrees[v], reverse=True),
        sorted(range(n), key=lambda v: degrees[v]),
        sorted(range(n), key=lambda v: (degrees[v], v), reverse=True),
        sorted(range(n), key=lambda v: (-degrees[v], v)),
        list(range(n)),
        list(reversed(range(n))),
    ]
    if degeneracy:
        static_orders.append(list(degeneracy))
        static_orders.append(list(reversed(degeneracy)))

    for order in static_orders:
        if time.perf_counter() >= deadline:
            break
        candidate = greedy_clique(neighbor_sets, order)
        if len(candidate) > len(best):
            best = candidate

    top_starts = sorted(range(n), key=lambda v: degrees[v], reverse=True)[: min(16, n)]
    for start in top_starts:
        if time.perf_counter() >= deadline:
            break
        candidate = greedy_clique_grasp(
            neighbor_sets,
            rng,
            start_vertex=start,
            neighbor_masks=neighbor_masks,
        )
        if len(candidate) > len(best):
            best = candidate

    while time.perf_counter() < deadline:
        if rng.random() < 0.5 and top_starts:
            start = rng.choice(top_starts)
            candidate = greedy_clique_grasp(
                neighbor_sets,
                rng,
                start_vertex=start,
                rcl_size=rng.randint(3, 6),
                neighbor_masks=neighbor_masks,
            )
        else:
            order = list(range(n))
            rng.shuffle(order)
            candidate = greedy_clique(neighbor_sets, order)
        if len(candidate) > len(best):
            best = candidate

    return best


def _swap_candidates(
    neighbor_sets: list[set[int]],
    base_set: set[int],
    n: int,
) -> list[int]:
    return [
        vertex
        for vertex in range(n)
        if vertex not in base_set and base_set.issubset(neighbor_sets[vertex])
    ]


def local_search(
    neighbor_sets: list[set[int]],
    clique: list[int],
    deadline: float,
    rng: random.Random,
) -> list[int]:
    """1-swap, (1,2)-add, (2,3)-add, and plateau moves."""
    best = list(clique)
    best_set = set(best)
    n = len(neighbor_sets)

    while time.perf_counter() < deadline:
        if not best:
            break

        remove_vertices: list[int]
        if rng.random() < 0.2 and len(best) >= 2:
            ranked = sorted(best, key=lambda v: len(neighbor_sets[v]))
            remove_vertices = ranked[:2]
        elif rng.random() < 0.75:
            remove_vertices = [min(best, key=lambda v: len(neighbor_sets[v]))]
        else:
            remove_vertices = [rng.choice(best)]

        reduced = best_set - set(remove_vertices)
        candidates = _swap_candidates(neighbor_sets, reduced, n)
        if not candidates:
            continue

        improved = False

        if len(remove_vertices) == 1:
            add_vertex = max(candidates, key=lambda v: len(neighbor_sets[v]))
            trial = sorted(reduced | {add_vertex})
            if len(trial) >= len(best):
                best = trial
                best_set = set(best)
                improved = True

            if not improved and len(best) >= 2 and rng.random() < 0.35:
                for i, u in enumerate(candidates):
                    for v in candidates[i + 1 :]:
                        if v in neighbor_sets[u]:
                            pair_trial = sorted(reduced | {u, v})
                            if len(pair_trial) > len(best):
                                best = pair_trial
                                best_set = set(best)
                                improved = True
                                break
                    if improved:
                        break

        if not improved and len(remove_vertices) == 2 and len(candidates) >= 3:
            for i, u in enumerate(candidates):
                for j in range(i + 1, len(candidates)):
                    v = candidates[j]
                    if v not in neighbor_sets[u]:
                        continue
                    for w in candidates[j + 1 :]:
                        if w in neighbor_sets[u] and w in neighbor_sets[v]:
                            triple = sorted(reduced | {u, v, w})
                            if len(triple) > len(best):
                                best = triple
                                best_set = set(best)
                                improved = True
                                break
                    if improved:
                        break
                if improved:
                    break

        if not improved and len(remove_vertices) == 1:
            add_vertex = rng.choice(candidates)
            trial = sorted(reduced | {add_vertex})
            if len(trial) == len(best):
                best = trial
                best_set = set(best)

    return best


def bitset_local_search(
    masks: list[int],
    n: int,
    init_clique: list[int],
    deadline: float,
    rng: random.Random,
    penalties: list[int] | None = None,
    restart_no_improve: int = 0,
) -> list[int]:
    """
    Incremental bitmask local search with penalty-guided escape (DLS-MC style).

    Grows the clique greedily; when maximal, attempts a (1,2)-swap, otherwise
    penalizes current members and drops one to diversify.
    """
    clique = list(init_clique)
    clique_mask = 0
    for vertex in clique:
        clique_mask |= 1 << vertex

    best = list(clique)
    best_size = len(clique)
    if penalties is None:
        penalties = [0] * n

    full_mask = (1 << n) - 1

    def add_set() -> int:
        mask = full_mask
        for vertex in clique:
            mask &= masks[vertex]
        return mask

    add_mask = add_set()
    no_improve = 0
    check_counter = 0

    while True:
        check_counter += 1
        if check_counter >= 256:
            check_counter = 0
            if time.perf_counter() >= deadline:
                break

        if add_mask:
            if rng.random() < 0.7:
                chosen = -1
                chosen_score = None
                for vertex in _iter_bits(add_mask):
                    score = (masks[vertex] & add_mask).bit_count() - penalties[vertex]
                    if chosen_score is None or score > chosen_score:
                        chosen_score = score
                        chosen = vertex
            else:
                candidates = list(_iter_bits(add_mask))
                chosen = rng.choice(candidates)

            clique.append(chosen)
            clique_mask |= 1 << chosen
            add_mask &= masks[chosen]

            if len(clique) > best_size:
                best = list(clique)
                best_size = len(clique)
                no_improve = 0
            continue

        no_improve += 1
        improved = False

        if clique:
            for _ in range(min(len(clique), 4)):
                pivot = rng.choice(clique)
                mask = full_mask
                for vertex in clique:
                    if vertex != pivot:
                        mask &= masks[vertex]
                mask &= ~clique_mask

                pair = None
                for first in _iter_bits(mask):
                    rest = masks[first] & mask & ~(1 << first)
                    if rest:
                        second = (rest & -rest).bit_length() - 1
                        pair = (first, second)
                        break

                if pair is not None:
                    first, second = pair
                    clique.remove(pivot)
                    clique_mask &= ~(1 << pivot)
                    clique.extend((first, second))
                    clique_mask |= (1 << first) | (1 << second)
                    add_mask = add_set()
                    if len(clique) > best_size:
                        best = list(clique)
                        best_size = len(clique)
                        no_improve = 0
                    improved = True
                    break

        if not improved and clique:
            for vertex in clique:
                penalties[vertex] += 1
            drop = max(clique, key=lambda u: (penalties[u], rng.random()))
            clique.remove(drop)
            clique_mask &= ~(1 << drop)
            add_mask = add_set()

        if restart_no_improve and no_improve >= restart_no_improve:
            clique = list(best)
            clique_mask = 0
            for vertex in clique:
                clique_mask |= 1 << vertex
            add_mask = add_set()
            no_improve = 0

    return best

File: crates/model_upgrade_core/src/heuristics.rs
use std::collections::HashSet;
use std::time::Instant;

use rand::prelude::*;
use rand::rngs::StdRng;

use crate::bitsets::{mask_degree, mask_intersection, BitMask, MAX_BITSET_NODES};

pub fn greedy_clique(neighbor_sets: &[HashSet<usize>], vertex_order: &[usize]) -> Vec<usize> {
    let mut clique: Vec<usize> = Vec::new();
    let mut clique_set: HashSet<usize> = HashSet::new();
    for &vertex in vertex_order {
        if clique_set.is_empty() || clique_set.is_subset(&neighbor_sets[vertex]) {
            clique.push(vertex);
            clique_set.insert(vertex);
        }
    }
    clique
}

pub fn greedy_clique_grasp(
    neighbor_sets: &[HashSet<usize>],
    rng: &mut StdRng,
    start_vertex: Option<usize>,
    rcl_size: usize,
    neighbor_masks: Option<&[BitMask]>,
) -> Vec<usize> {
    let n = neighbor_sets.len();
    let use_masks = neighbor_masks.is_some() && n <= MAX_BITSET_NODES;

    let mut clique: Vec<usize> = Vec::new();
    if let Some(sv) = start_vertex {
        clique.push(sv);
    }

    if use_masks {
        let masks = neighbor_masks.unwrap();
        let mut candidates_mask = if clique.is_empty() {
            BitMask::all(n)
        } else {
            let refs: Vec<BitMask> = clique.iter().map(|&v| masks[v]).collect();
            let mut cm = mask_intersection(&refs);
            for &v in &clique {
                cm.clear(v);
            }
            cm
        };

        while !candidates_mask.is_empty() {
            let candidates = candidates_mask.vertices();
            let mut degrees: Vec<(usize, u32)> = candidates
                .iter()
                .map(|&v| (v, mask_degree(masks[v])))
                .collect();
            degrees.sort_by(|a, b| b.1.cmp(&a.1));
            let top_len = rcl_size.min(degrees.len());
            let top: Vec<usize> = degrees[..top_len].iter().map(|(v, _)| *v).collect();
            let chosen = *top.choose(rng).unwrap();
            clique.push(chosen);
            candidates_mask.and_assign(&masks[chosen]);
            candidates_mask.clear(chosen);
        }
        return clique;
    }

    let mut clique_set: HashSet<usize> = clique.iter().copied().collect();
    let mut candidates: HashSet<usize> = if clique_set.is_empty() {
        (0..n).collect()
    } else {
        clique
            .iter()
            .map(|&v| neighbor_sets[v].clone())
            .reduce(|a, b| a.intersection(&b).copied().collect())
            .unwrap_or_default()
    };

    while !candidates.is_empty() {
        let mut ranked: Vec<usize> = candidates.iter().copied().collect();
        ranked.sort_by_key(|&v| std::cmp::Reverse(neighbor_sets[v].len()));
        let top_len = rcl_size.min(ranked.len());
        let chosen = ranked[rng.random_range(0..top_len)];
        clique.push(chosen);
        clique_set.insert(chosen);
        candidates = candidates
            .intersection(&neighbor_sets[chosen])
            .copied()
            .collect();
        candidates = candidates
            .difference(&clique_set)
            .copied()
            .collect();
    }

    clique
}

pub fn random_restarts(
    neighbor_sets: &[HashSet<usize>],
    deadline: Instant,
    rng: &mut StdRng,
    degeneracy: Option<&[usize]>,
    neighbor_masks: Option<&[BitMask]>,
) -> Vec<usize> {
    let n = neighbor_sets.len();
    let degrees: Vec<usize> = neighbor_sets.iter().map(|s| s.len()).collect();
    let mut best: Vec<usize> = Vec::new();

    let mut static_orders: Vec<Vec<usize>> = vec![
        {
            let mut o: Vec<usize> = (0..n).collect();
            o.sort_by_key(|&v| std::cmp::Reverse(degrees[v]));
            o
        },
        {
            let mut o: Vec<usize> = (0..n).collect();
            o.sort_by_key(|&v| degrees[v]);
            o
        },
        {
            let mut o: Vec<usize> = (0..n).collect();
            o.sort_by(|&a, &b| (degrees[b], b).cmp(&(degrees[a], a)));
            o
        },
        {
            let mut o: Vec<usize> = (0..n).collect();
            o.sort_by(|&a, &b| (degrees[a], a).cmp(&(degrees[b], b)));
            o
        },
        (0..n).collect(),
        (0..n).rev().collect(),
    ];

    if let Some(degen) = degeneracy {
        static_orders.push(degen.to_vec());
        static_orders.push(degen.iter().rev().copied().collect());
    }

    for order in &static_orders {
        if Instant::now() >= deadline {
            break;
        }
        let candidate = greedy_clique(neighbor_sets, order);
        if candidate.len() > best.len() {
            best = candidate;
        }
    }

    let mut top_starts: Vec<usize> = (0..n).collect();
    top_starts.sort_by_key(|&v| std::cmp::Reverse(degrees[v]));
    top_starts.truncate(16.min(n));

    for &start in &top_starts {
        if Instant::now() >= deadline {
            break;
        }
        let candidate = greedy_clique_grasp(
            neighbor_sets,
            rng,
            Some(start),
            4,
            neighbor_masks,
        );
        if candidate.len() > best.len() {
            best = candidate;
        }
    }

    while Instant::now() < deadline {
        let candidate = if rng.random_bool(0.5) && !top_starts.is_empty() {
            let start = *top_starts.choose(rng).unwrap();
            let rcl = rng.random_range(3..=6);
            greedy_clique_grasp(neighbor_sets, rng, Some(start), rcl, neighbor_masks)
        } else {
            let mut order: Vec<usize> = (0..n).collect();
            order.shuffle(rng);
            greedy_clique(neighbor_sets, &order)
        };
        if candidate.len() > best.len() {
            best = candidate;
        }
    }

    best
}

fn swap_candidates(
    neighbor_sets: &[HashSet<usize>],
    base_set: &HashSet<usize>,
    n: usize,
) -> Vec<usize> {
    (0..n)
        .filter(|&vertex| {
            !base_set.contains(&vertex) && base_set.is_subset(&neighbor_sets[vertex])
        })
        .collect()
}

pub fn local_search(
    neighbor_sets: &[HashSet<usize>],
    clique: &[usize],
    deadline: Instant,
    rng: &mut StdRng,
) -> Vec<usize> {
    let mut best: Vec<usize> = clique.to_vec();
    let mut best_set: HashSet<usize> = best.iter().copied().collect();
    let n = neighbor_sets.len();

    while Instant::now() < deadline {
        if best.is_empty() {
            break;
        }

        let remove_vertices: Vec<usize> = if rng.random_bool(0.2) && best.len() >= 2 {
            let mut ranked = best.clone();
            ranked.sort_by_key(|&v| neighbor_sets[v].len());
            ranked[..2].to_vec()
        } else if rng.random_bool(0.75) {
            let v = *best
                .iter()
                .min_by_key(|&&v| neighbor_sets[v].len())
                .unwrap();
            vec![v]
        } else {
            vec![*best.choose(rng).unwrap()]
        };

        let reduced: HashSet<usize> = best_set
            .difference(&remove_vertices.iter().copied().collect())
            .copied()
            .collect();
        let candidates = swap_candidates(neighbor_sets, &reduced, n);
        if candidates.is_empty() {
            continue;
        }

        let mut improved = false;

        if remove_vertices.len() == 1 {
            let add_vertex = *candidates
                .iter()
                .max_by_key(|&&v| neighbor_sets[v].len())
                .unwrap();
            let mut trial: Vec<usize> = reduced.iter().copied().collect();
            trial.push(add_vertex);
            trial.sort_unstable();
            if trial.len() >= best.len() {
                best = trial;
                best_set = best.iter().copied().collect();
                improved = true;
            }

            if !improved && best.len() >= 2 && rng.random_bool(0.35) {
                'outer: for (i, &u) in candidates.iter().enumerate() {
                    for &v in &candidates[i + 1..] {
                        if neighbor_sets[u].contains(&v) {
                            let mut pair_trial: Vec<usize> = reduced.iter().copied().collect();
                            pair_trial.push(u);
                            pair_trial.push(v);
                            pair_trial.sort_unstable();
                            if pair_trial.len() > best.len() {
                                best = pair_trial;
                                best_set = best.iter().copied().collect();
                                improved = true;
                                break 'outer;
                            }
                        }
                    }
                }
            }
        }

        if !improved && remove_vertices.len() == 2 && candidates.len() >= 3 {
            'outer2: for (i, &u) in candidates.iter().enumerate() {
                for (j, &v) in candidates[i + 1..].iter().enumerate() {
                    let j = i + 1 + j;
                    if !neighbor_sets[u].contains(&v) {
                        continue;
                    }
                    for &w in &candidates[j + 1..] {
                        if neighbor_sets[u].contains(&w) && neighbor_sets[v].contains(&w) {
                            let mut triple: Vec<usize> = reduced.iter().copied().collect();
                            triple.extend([u, v, w]);
                            triple.sort_unstable();
                            if triple.len() > best.len() {
                                best = triple;
                                best_set = best.iter().copied().collect();
                                improved = true;
                                break 'outer2;
                            }
                        }
                    }
                }
            }
        }

        if !improved && remove_vertices.len() == 1 {
            let add_vertex = *candidates.choose(rng).unwrap();
            let mut trial: Vec<usize> = reduced.iter().copied().collect();
            trial.push(add_vertex);
            trial.sort_unstable();
            if trial.len() == best.len() {
                best = trial;
                best_set = best.iter().copied().collect();
            }
        }
    }

    best
}

pub fn bitset_local_search(
    masks: &[BitMask],
    n: usize,
    init_clique: &[usize],
    deadline: Instant,
    rng: &mut StdRng,
    penalties: &mut [i32],
    restart_no_improve: usize,
) -> Vec<usize> {
    let mut clique: Vec<usize> = init_clique.to_vec();
    let mut clique_mask = BitMask::empty();
    for &vertex in &clique {
        clique_mask.set(vertex);
    }

    let mut best = clique.clone();
    let mut best_size = clique.len();
    let full_mask = BitMask::all(n);

    let mut add_mask = {
        let mut mask = full_mask;
        for &vertex in &clique {
            mask.and_assign(&masks[vertex]);
        }
        mask
    };

    let mut no_improve = 0usize;
    let mut check_counter = 0u32;

    loop {
        check_counter += 1;
        if check_counter >= 256 {
            check_counter = 0;
            if Instant::now() >= deadline {
                break;
            }
        }

        if !add_mask.is_empty() {
            let chosen = if rng.random_bool(0.7) {
                let mut best_vertex = 0usize;
                let mut best_score = i32::MIN;
                let mut first = true;
                for vertex in add_mask.iter_bits() {
                    let score = (masks[vertex].and(add_mask).count() as i32) - penalties[vertex];
                    if first || score > best_score {
                        best_score = score;
                        best_vertex = vertex;
                        first = false;
                    }
                }
                best_vertex
            } else {
                let candidates: Vec<usize> = add_mask.vertices();
                *candidates.choose(rng).unwrap()
            };

            clique.push(chosen);
            clique_mask.set(chosen);
            add_mask.and_assign(&masks[chosen]);

            if clique.len() > best_size {
                best = clique.clone();
                best_size = clique.len();
                no_improve = 0;
            }
            continue;
        }

        no_improve += 1;
        let mut improved = false;

        if !clique.is_empty() {
            for _ in 0..clique.len().min(4) {
                let pivot = *clique.choose(rng).unwrap();
                let mut mask = full_mask;
                for &vertex in &clique {
                    if vertex != pivot {
                        mask.and_assign(&masks[vertex]);
                    }
                }
                mask.and_not_assign(&clique_mask);

                let mut pair: Option<(usize, usize)> = None;
                for first in mask.iter_bits() {
                    let mut rest = masks[first].and(mask);
                    rest.clear(first);
                    if let Some(second) = rest.lowest_bit() {
                        pair = Some((first, second));
                        break;
                    }
                }

                if let Some((first, second)) = pair {
                    clique.retain(|&v| v != pivot);
                    clique_mask.clear(pivot);
                    clique.push(first);
                    clique.push(second);
                    clique_mask.set(first);
                    clique_mask.set(second);
                    add_mask = {
                        let mut m = full_mask;
                        for &vertex in &clique {
                            m.and_assign(&masks[vertex]);
                        }
                        m
                    };
                    if clique.len() > best_size {
                        best = clique.clone();
                        best_size = clique.len();
                        no_improve = 0;
                    }
                    improved = true;
                    break;
                }
            }
        }

        if !improved && !clique.is_empty() {
            for &vertex in &clique {
                penalties[vertex] += 1;
            }
            let drop = *clique
                .iter()
                .max_by(|&&a, &&b| {
                    penalties[a]
                        .cmp(&penalties[b])
                        .then_with(|| rng.random::<u32>().cmp(&rng.random()))
                })
                .unwrap();
            clique.retain(|&v| v != drop);
            clique_mask.clear(drop);
            add_mask = {
                let mut m = full_mask;
                for &vertex in &clique {
                    m.and_assign(&masks[vertex]);
                }
                m
            };
        }

        if restart_no_improve > 0 && no_improve >= restart_no_improve {
            clique = best.clone();
            clique_mask = BitMask::empty();
            for &vertex in &clique {
                clique_mask.set(vertex);
            }
            add_mask = {
                let mut m = full_mask;
                for &vertex in &clique {
                    m.and_assign(&masks[vertex]);
                }
                m
            };
            no_improve = 0;
        }
    }

    best
}
